circuits/circuits/send.circom

```diff
@@ -9,42 +9,85 @@ template Send(levels) {
     signal input recipient;
     signal input initial_root;
 
+    signal input leaf_coins[2];
+    signal input pathElements[levels];
+    signal input pathIndices[levels];
+
+    signal input highest_coin_to_send; // sending [leaf_coins[0], highest_coin_to_send]
+    signal input signature;
+
+    signal input pathElementsForZero[levels];
+    signal input pathIndicesForZero[levels];
+
     signal output new_root;
 
-    // make sure it's a send request by checking sender and recipient is non-zero
-    signal is_sender_zero <== IsZero(in <== sender);
-    is_sender_zero === 0;
-    signal is_recipient_zero <== IsZero(in <== recipient);
-    is_recipient_zero === 0;
+    {
+        // make sure it's a send request by checking sender and recipient is non-zero
+        signal is_sender_zero <== IsZero()(in <== sender);
+        is_sender_zero === 0;
+        signal is_recipient_zero <== IsZero()(in <== recipient);
+        is_recipient_zero === 0;
+    }
 
     // To process a transaction we:
     // - Delete the old leaf
-    // - Insert up to 3 new leaves
-    // - Merge up to 3 adjacent new leaves
+    // - Insert 2 new leaves
     //
     // We then output the new root from the circuit
-    component transaction_is_valid = Validate();
-
-    signal next_root = 0; // TODO
+    signal is_transaction_valid <== Validate(levels)(
+        initial_root <== initial_root,
+        highest_coin_to_send <== highest_coin_to_send,
+        signature <== signature,
+        owner <== sender,
+        leaf_coins <== leaf_coins,
+        pathElements <== pathElements,
+        pathIndices <== pathIndices
+    );
+
+    // determine the new leaf corresponding to sender.
+    // If sending all coins, we replace it with 0; otherwise we replace it with the remaining coins.
+    signal is_send_all <== IsEqual()(in <== [highest_coin_to_send, leaf_coins[1]]);
+    signal sender_coin_leaf <== Poseidon(3)(inputs <== [sender, highest_coin_to_send+1, leaf_coins[1]]);
```

Comment:
Man, this syntax is *so good* compared to what circom used to be. Can't get over it.

Reply:
makes you want to just stare at the code in awe 🫂